wsd: save errno before invoking any other functions

Ashod opened the pull request:
Most C and Posix API clobber errno. By failing to save
it immediately after invoking an API we risk simply
reporting the result of an arbitrary subsequent API call.

This changes LOG_SYS and LOG_SFL to take errno explicitly.
This is necessary because sometimes logging is not done
immediately after calling the function for which we
want to report errno. Similarly, both of these log
macros need to save errno before calling any functions.
This is necessary in case the argument is in fact errno
and not a copy.

Finally, both of these log macros append errno to the
log message, so there is little point in ending the
messages with a period.

Change-Id: Iecc656f67115fec78b65cad4e7c17a17623ecf43
Signed-off-by: Ashod Nakashian <ashod.nakashian@collabora.co.uk>

mmeeks reviewed (changes requested):
Surely LOG_SYS - can simply cache errno in it's first line itself into a temporary variable if you are worried ? it is after all a macro ;-) lets not do this widespread code-change. If there is somewhere a macro that needs to pass a custom errno in then we can have LOG_SYS_ERRNO instead to explicitly pass it I suppose.

Ashod commented:
> Surely LOG_SYS - can simply cache errno in it's first line itself into a temporary variable if you are worried ? it is after all a macro ;-) 

That's exactly what LOG_SYS does now, with my patch. That was one error fixed.

> lets not do this widespread code-change. If there is somewhere a macro that needs to pass a custom errno in then we can have LOG_SYS_ERRNO instead to explicitly pass it I suppose.

Adding another macro is easy, but the problem is that apparently it's equally easy to miss the tiny detail that errno is implicit in LOG_SYS and clobber it by inserting a function call before LOG_SYS.

I don't mind adding LOG_SYS_ERRNO, but it won't fix the problem and this will resurface rather soon.

Ashod commented:
@mmeeks, reverted the changes to LOG_SYS. I still think using errno implicitly is poor design (if the author may say so himself) and leads to wrong usage (as I fixed in a number of places in this PR).

But it's a shorter patch to review, so there is that win.

mmeeks reviewed (approved):
Thanks ! =)
